ops/model.py

```diff
@@ -1771,6 +1782,71 @@ def remote_model(self) -> RemoteModel:
         return self._remote_model
 
+    def load_data(
+        self,
+        cls: type[_InterfaceType],
+        app_or_unit: Union[Unit, Application],
+        *args: Any,
+        decoder: Callable[[Any], Any] | None = None,
+        encoder: Callable[[Any], Any] | None = None,
+        **kwargs: Any,
+    ) -> _InterfaceType:
+        """Load the data for this relation into an instance of a databag class.
+
+        The object will be instantiated with the data from the relation databag.
+        If the app or unit data has already been loaded into an object, then the
+        same object is returned (and the decoder, encoder and positional and
+        keyword arguments passed in this call are not used).
+
+        Any additional positional or keyword arguments will be passed through to
+        the databag class.
+
+        Args:
+            cls: A class that inherits from :class:`ops.DatabagBase`.
+            app_or_unit: The databag to load. This can be either a :class:`Unit`
+                or :class:`Application` instance.
+            decoder: An optional callable that will be used to decode each field
+                before loading into the class. If not provided, json.loads will
+                be used.
+            encoder: An optional callable that will be used to encode each field
+                before passing to Juju. If not provided, json.dumps will be
+                used.
+            args: positional arguments to pass through to the databag class.
+            kwargs: keyword arguments to pass through to the databag class.
+
+        Returns:
+            An instance of the databag class with the current relation data values.
+
+        Raises:
+            TypeError: if the same app or unit data is already loaded into a
+                different databag class.
+            InvalidSchemaError: if the databag class cannot be instantiated with the
+                provided data.
+        """
+        # If we have already created an object connected to the databag, then
+        # we need to return the same object, so that users can't make conflicting
+        # changes that we would need to resolve when sending the data to Juju.
+        if (self.id, app_or_unit) in self._cache._databag_obj_cache:
+            obj = self._cache._databag_obj_cache[self.id, app_or_unit][0]
+            if type(obj) is not cls:
+                raise TypeError(
+                    f'Cannot load data for {app_or_unit} into {cls.__name__}, '
+                    f'it is already loaded into {type(obj).__name__}'
+                )
+            return obj
+        data = copy.deepcopy(kwargs)
+        if decoder is None:
+            decoder = json.loads
+        content = {k: decoder(v) for k, v in self.data[app_or_unit].items()}
+        data.update(**content)
+        try:
+            obj = cls(*args, **data)
+        except ValueError as e:
+            self._backend.status_set('waiting', f'Error in relation data: {e}')
```

Comment:
This should maybe be done at the end when _main catches the exception, so that if the charm catches it the status isn't set (which would send through immediately to Juju, and might override a status the charm already set).